#Get chromosome information from hapmap data
# Documentation using roxygen2
#' Get chromosome information from hapmap data
#' @Description Get chromosome information from hapmap data
#' @param hapData a hapmap data
#' @return a data frame of chromosome information
#' @export
getChrInfoFromHap<-function(hapData)
{
  # select columns chrom, and pos
  hapData<-hapData[,c("chrom","pos")]
  # get the min and max position for each chromosome
  hapData<-aggregate(pos~chrom, hapData, range)
  hapData<-as.matrix(hapData)
  # rename the columns
  colnames(hapData)<-c("chr","start","end")
  # convert to data frame
  hapData<-as.data.frame(hapData)
  hapData$start<-as.numeric(hapData$start)
  hapData$end<-as.numeric(hapData$end)
  # add the length
  hapData$length<-hapData$end-hapData$start
  #sort chromosomes by numbers inside the name
  hapData$order<-hapData$chr
  # remove all characters from the chromosome name
  hapData$order<-gsub("[^0-9]", "", hapData$order)
  # convert to numeric
  hapData$order<-as.numeric(hapData$order)
  # sort by order
  hapData<-hapData[order(hapData$order),]
  #make the chr column as factor and add the levels 
  hapData$chr<-factor(hapData$chr,levels=hapData$chr)
  # remove the order column
  hapData$order<-NULL
  # add description
  description(hapData)<-"chromosome information"
  description(hapData$chr)<-"chromosome name"
  description(hapData$start)<-"start position"
  description(hapData$end)<-"end position"
  description(hapData$length)<-"length"
  return(hapData)
}

#' Get chromosome information from SNPINFO file
#' @Description Get chromosome information from SNPINFO file
#' @param SNPINFO a SNPINFO file
#' @return a data frame of chromosome information
#' @export
getChrInfoFromSNPINFO<-function(SNPINFO)
{
  # The difference between this function and the one above is that this function takes the SNPINFO file as input that was generated by the GWAS file
  # If the SNPs are not in the gwas file, then the SNPINFO file will not have the chromosome information
  SNPINFO<-SNPINFO[,c("chr","ps")]
  SNPINFO<-aggregate(ps~chr, SNPINFO, range)
  SNPINFO<-as.matrix(SNPINFO)
  colnames(SNPINFO)<-c("chr","start","end")
  SNPINFO<-as.data.frame(SNPINFO)
  SNPINFO$start<-as.numeric(SNPINFO$start)
  SNPINFO$end<-as.numeric(SNPINFO$end)
  SNPINFO$length<-SNPINFO$end-SNPINFO$start
  SNPINFO$order<-SNPINFO$chr
  SNPINFO$order<-gsub("[^0-9]", "", SNPINFO$order)
  SNPINFO$order<-as.numeric(SNPINFO$order)
  SNPINFO<-SNPINFO[order(SNPINFO$order),]
  SNPINFO$chr<-factor(SNPINFO$chr,levels=SNPINFO$chr)
  SNPINFO$order<-NULL
  return(SNPINFO)
}

# Documentation using roxygen2
#' Get chromosome information from GWAS file
#' @Description Get chromosome information from GWAS file
#' @param GWAS a GWAS file
#' @return a data frame of chromosome information
#' @export
getChrInfoFromSNPGWAS<-function(GWAS)
{
  SNPINFO <- getMarkersInfoFromGWAS(GWAS)
  ChrInfo <- getChrInfoFromSNPINFO(SNPINFO)
  return(ChrInfo)
}
